### PR TITLE
feat(fetch): support following redirect responses

### LINKED ISSUE #626: Support following fetch redirect responses
There are instances where redirect responses to fetch requests aren't being followed:

- https://github.com/mswjs/msw/issues/2266
- https://github.com/nock/nock/issues/2755

## Caution

Before fixing this, we need to write compliance tests for different request modules in accordance to how they follow or _don't_ follow redirects. Note that not all request modules may have redirect following built-in. For example, I believe the `http` module doesn't do that, thus the existence of designated third-party modules just for that functionality. 

## Roadmap

1. Add failing compliance tests for following redirect responses (can start from `fetch`). A good start: https://github.com/mswjs/interceptors/pull/581
1. Analyze existing solutions to learn how they implement redirect response following. Use Undici as an example, I believe it has that built-in. 
1. Implement the redirect following.

### PR DESCRIPTION
**Pull request by kettanaito**
> [!WARNING]
> Marking this as a breaking (feat) release because mocked redirect responses will now be followed by default (previously, returned as-is). The previous behavior was faulty and not spec-compliant, but I don't want to break consumers once again. 

- Fixes #626 
- Fixes https://github.com/mswjs/msw/issues/2266
- Related to #575 

The redirect response handling is taken from Undici almost as-is because it doesn't export its `web/fetch/utils` as a typed module (it's untyped CJS). Luckily, Undici follows the spec in those implementations so they are unlikely to change. I'm also omitting things that are exclusive to Undici, such as accessing `request.body.source`. 

## Relevant resources

- https://github.com/nodejs/undici/blob/a6dac3149c505b58d2e6d068b97f4dc993da55f0/lib/web/fetch/index.js#L1173
- https://github.com/nodejs/undici/blob/a6dac3149c505b58d2e6d068b97f4dc993da55f0/lib/web/fetch/index.js#L1210

### CODE REVIEW

File: test/modules/fetch/compliance/fetch-follow-redirects.test.ts [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+afterEach(() => {
+  interceptor.removeAllListeners()
+})
+
+afterAll(async () => {
+  interceptor.dispose()
+  await httpServer.close()
+})
+
+it('follows a bypassed redirect response', async () => {
```
Comment (kettanaito) [inline test/modules/fetch/compliance/fetch-follow-redirects.test.ts:29]: Added the tests for all scenarios we can reproduce from Undici.

File: src/interceptors/fetch/utils/followRedirect.ts [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+  'content-type',
+  'content-length',
+]
+
+const kRedirectCount = Symbol('kRedirectCount')
+
+/**
+ * @see https://github.com/nodejs/undici/blob/a6dac3149c505b58d2e6d068b97f4dc993da55f0/lib/web/fetch/index.js#L1210
+ */
+export async function followFetchRedirect(
```
Comment (kettanaito) [inline src/interceptors/fetch/utils/followRedirect.ts:16]: Implementation of following redirects.

File: src/interceptors/fetch/index.ts [hunk 3]
```diff
@@ -68,6 +71,33 @@ export class FetchInterceptor extends Interceptor<HttpRequestEventMap> {
             response,
           })
 
+          /**
+           * Undici's handling of following redirect responses.
+           * Treat the "manual" redirect mode as a regular mocked response.
+           * This way, the client can manually follow the redirect it receives.
+           * @see https://github.com/nodejs/undici/blob/a6dac3149c505b58d2e6d068b97f4dc993da55f0/lib/web/fetch/index.js#L1173
+           */
+          if (RESPONSE_STATUS_CODES_WITH_REDIRECT.has(response.status)) {
```
Comment (kettanaito) [inline src/interceptors/fetch/index.ts:80]: Fix: handle redirect responses in a special way.

**Comment by kettanaito** [comment]
## Released: v0.35.0 🎉

This has been released in v0.35.0!

- 📄 [**Release notes**](https://github.com/mswjs/interceptors/releases/tag/v0.35.0)
- 📦 [npm package](https://www.npmjs.com/package/@mswjs/interceptors/v/0.35.0)

Make sure to always update to the latest version (`npm i @mswjs/interceptors@latest`) to get the newest features and bug fixes.

---

_Predictable release automation by [@ossjs/release](https://github.com/ossjs/release)_.

**Comment by mikicho** [comment]
Amazing PR!

**Pull request merged**